src/core/layertree/qgslayertreemodellegendnode.cpp

```diff
@@ -530,34 +548,65 @@ void QgsSymbolLegendNode::updateLabel()
       qlonglong count = vl->featureCount( mItem.ruleKey() );
       mLabel += QStringLiteral( " [%1]" ).arg( count != -1 ? QLocale().toString( count ) : tr( "N/A" ) );
     }
+    else if ( vl )
+    {
+      mLabel = evaluateLabel( mLabel, vl );
+    }
   }
-
   emit dataChanged();
 }
 
+QgsExpressionContext QgsSymbolLegendNode::createExpressionContext() const
+{
+  QgsExpressionContext context; //= QgsLayoutItem::createExpressionContext();
 
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
 
-// -------------------------------------------------------------------------
+  // TODO: ADD NECESSARY SCOPES ??
+  context.appendScope( vl->createExpressionContextScope() );
 
 
-QgsSimpleLegendNode::QgsSimpleLegendNode( QgsLayerTreeLayer *nodeLayer, const QString &label, const QIcon &icon, QObject *parent, const QString &key )
-  : QgsLayerTreeModelLegendNode( nodeLayer, parent )
-  , mLabel( label )
-  , mIcon( icon )
-  , mKey( key )
-{
+  QgsExpressionContextScope *scope = new QgsExpressionContextScope( tr( "Symbol scope" ) );
+
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_label" ), textOnSymbolLabel(), true ) );
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_id" ), mItem.ruleKey(), true ) );
+  QgsVectorLayerFeatureCounter *counter = vl->countSymbolFeatures() ;
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_count" ), QVariant::fromValue( counter->featureCount( mItem.ruleKey() ) ), true ) );
+
+
+  QVariantList featureIds;
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_feature_ids" ), featureIds, true ) );
+
+  QgsFeatureIds fids = counter->featureIds( mItem.ruleKey() );
+
+  featureIds.reserve( fids.count() );
+  for ( QgsFeatureId fid : fids )
+  {
+    featureIds << static_cast<qint64>( fid );
+  }
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_feature_ids" ), featureIds, true ) );
+
+
+  context.appendScope( scope );
+
+  return context;
 }
 
-QVariant QgsSimpleLegendNode::data( int role ) const
+QString QgsSymbolLegendNode::evaluateLabel( QString label, QgsVectorLayer *vl ) const
 {
-  if ( role == Qt::DisplayRole || role == Qt::EditRole )
-    return mUserLabel.isEmpty() ? mLabel : mUserLabel;
-  else if ( role == Qt::DecorationRole )
-    return mIcon;
-  else if ( role == RuleKeyRole && !mKey.isEmpty() )
-    return mKey;
+  QgsExpressionContext context;
+  if ( mLayerNode->layer()->type() == VectorLayer )
```

Comment:
I'm not sure about this, can anyone help ( @nyalldawson , @m-kuhn ), I just want to ensure that the source layer is a vector, using vl would mean that the source layer is already casted as a layer and that is not viable.

Do I need to add QgsMapLayer.h in the inclue , if so what would be the syntax to check the returned type, I assume that I should use QgsMapLayer::LayerType::VectorLayer ?

I need to do this since my previous wms check was useless since the wms sources are spatialite for the test and this is what lead to the segfault.